test/src/main/java/org/apache/accumulo/test/fate/zookeeper/ZooLockIT.java

```diff
@@ -306,6 +363,245 @@ public void testUnexpectedEvent() throws Exception {
+          lockCompletedLatch.countDown();
+          unlockLatch.await();
+          zl.unlock();
+        }
+      } catch (IOException | InterruptedException | KeeperException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public String toString() {
+      return "LockWorker [name=" + name + ", holdsLock()=" + holdsLock() + "]";
+    }
+
+  }
+
+  private int parseLockWorkerName(String child) {
+    if (child.startsWith("zlock#00000000-0000-0000-0000-000000000000#")) {
+      return 0;
+    } else if (child.startsWith("zlock#00000000-0000-0000-0000-111111111111#")) {
+      return 1;
+    } else if (child.startsWith("zlock#00000000-0000-0000-0000-222222222222#")) {
+      return 2;
+    } else if (child.startsWith("zlock#00000000-0000-0000-0000-333333333333#")) {
+      return 3;
+    } else {
+      return -1;
+    }
+  }
+
+  @Test(timeout = 60000)
+  public void testLockParallel() throws Exception {
+    String parent = "/zlParallel";
+
+    ConnectedWatcher watcher = new ConnectedWatcher();
+    try (ZooKeeperWrapper zk = new ZooKeeperWrapper(getCluster().getZooKeepers(), 30000, watcher)) {
+      zk.addAuthInfo("digest", "accumulo:secret".getBytes(UTF_8));
+
+      while (!watcher.isConnected()) {
+        Thread.sleep(50);
+      }
+      // Create the parent node
+      zk.createOnce(parent, new byte[0], ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+      int numWorkers = 4;
+      final CountDownLatch getLockLatch = new CountDownLatch(numWorkers);
+      final CountDownLatch lockFinishedLatch = new CountDownLatch(numWorkers);
+      final List<LockWorker> workers = new ArrayList<>(numWorkers);
+      final List<Thread> threads = new ArrayList<>(numWorkers);
+      for (int i = 0; i < numWorkers; i++) {
+        String name =
+            "zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#".replaceAll("A", Integer.toString(i));
+        LockWorker w = new LockWorker(parent, name, getLockLatch, lockFinishedLatch);
+        Thread t = new Thread(w);
+        workers.add(w);
+        threads.add(t);
+        t.start();
+      }
+
+      getLockLatch.await(); // Threads compete for lock
+      lockFinishedLatch.await(); // Threads lock logic complete
+
+      for (int i = 4; i > 0; i--) {
+        List<String> children = zk.getChildren(parent, false);
+        ZooLock.sortChildrenByLockPrefix(children);
+        assertEquals(i, children.size());
+        String first = children.get(0);
+        int workerWithLock = parseLockWorkerName(first);
+        LockWorker worker = workers.get(workerWithLock);
+        assertTrue(worker.holdsLock());
+        workers.forEach(w -> {
+          if (w != worker) {
+            assertFalse(w.holdsLock());
+          }
+        });
+        worker.unlock();
+        Thread.sleep(500); // wait for node to be deleted and lock to be reacquired.
+      }
+
+      workers.forEach(w -> assertFalse(w.holdsLock()));
+
```

Comment:
Does it make sense to also check the following here?

```java
assertEquals(0, zk.getChildren(parent, false));
```

Reply:
added to test